### PR TITLE
Make BatchLoader work with the latest GraphQL versions

### PR DESCRIPTION
**Pull request by exAspArk**
TL;DR

* If you use `graphql` gem version `1.8.6` or lower, you can continue using `BatchLoader`.
* If you use `graphql` gem version `1.8.7` or greater, you need to replace `BatchLoader.for` with `BatchLoader::GraphQL.for` to make it work within your GraphQL resolvers (methods).

---

Since version [v1.0.0](https://github.com/exAspArk/batch-loader/blob/master/CHANGELOG.md#v100--2017-08-21), `BatchLoader` behaves as a lazy object. It only starts batching when it needs to resolve the values:

```ruby
batch_loader = BatchLoader.for(1).batch(default_value: :foo) { puts "batching..." }
# => #<BatchLoader:0x140623647733240>

batch_loader.class
# batching...
# => Symbol

puts batch_loader
# foo
# => nil
```

At the same time, `graphql-ruby` gem implements a general adapter interface for different [lazy execution](https://github.com/rmosolgo/graphql-ruby/blob/75c3529177474c8e65fd175f0c36fdc2fac185b1/guides/schema/lazy_execution.md) mechanisms. This interface relies on an instance [class](https://github.com/rmosolgo/graphql-ruby/blob/75c3529177474c8e65fd175f0c36fdc2fac185b1/lib/graphql/execution/lazy/lazy_method_map.rb#L36) method to detect whether an object is lazy or not. Unfortunately, since calling `batch_loader.class` tries to resolve the values, it is being resolved with N+1 queries by `graphql-ruby` gem itself.

```ruby
batch_loader.class
# batching...
# => Symbol
```

I [suggested](https://github.com/rmosolgo/graphql-ruby/pull/1784#issuecomment-417736649) a few other potentially more flexible solutions for `graphql-ruby` to detect lazy objects such as duck typing or using explicit arguments. But it looks like it won't be implemented. To fix the issue `BatchLoader` started wrapping `BatchLoader` objects with PORO (plain old ruby objects) by using `graphql-ruby` [instrumentation](https://github.com/rmosolgo/graphql-ruby/blob/75c3529177474c8e65fd175f0c36fdc2fac185b1/guides/fields/instrumentation.md).

This simple `BatchLoader` object wrapping stopped working since `graphql-ruby` version `1.8.7`. This version introduced a breaking change and [started detecting](https://github.com/rmosolgo/graphql-ruby/issues/1778#issuecomment-413331301) lazy objects before the instrumentation. 

In order to fix the compatibility again, this PR introduces `BatchLoader::GraphQL` - a simple PORO with extra 20 lines of code to wrap `BatchLoader` without relying on `graphql-ruby` instrumentation. For example:

```ruby
class PostType < GraphQL::Schema::Object
  field :user, UserType, null: false

  def user
    BatchLoader::GraphQL.for(object.user_id).batch do |user_ids, loader|
      User.where(id: user_ids).each { |user| loader.call(user.id, user) }
    end
  end
end
```

---

Related to https://github.com/rmosolgo/graphql-ruby/issues/1778. Fixes https://github.com/exAspArk/batch-loader/issues/22, https://github.com/exAspArk/batch-loader/issues/26, https://github.com/exAspArk/batch-loader/issues/30.

### CODE REVIEW

**Comment by maletor** [comment]
This created a regression where resolvers that do not *return* a `BatchLoader::GraphQL` fail to call `BatchLoader::GraphQL` calls that the resolver uses.
